## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
basic_statistics = df.describe()
print(basic_statistics)


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [4]:
# What three item categories had the most entries?
most_entries = df['category'].value_counts().head(3)
print(most_entries)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [5]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax

cat_specific = df[df['category'] == top_category]

top_subcategory = cat_specific['subcategory'].value_counts().idxmax

In [7]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head()
print(top_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [8]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()
print(top_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [10]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = df['client_id'].value_counts().idxmax()

client_specific = df[df['client_id'] == top_client_id]

all_entries = client_specific['qty'].sum()
print(all_entries)

64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [21]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df['subtotal'])

0        115164.00
1           523.95
2           527.28
3          1056.18
4          3902.00
           ...    
54634      2743.29
54635      9709.73
54636     31188.50
54637       165.76
54638      3103.31
Name: subtotal, Length: 54639, dtype: float64


In [20]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: x * 7 if x > 50
                                             else x * 10)
print(df['shipping_price'])



0         75.0
1         14.9
2         16.8
3         12.3
4        464.3
         ...  
54634     22.5
54635    117.0
54636     41.6
54637    180.4
54638     20.7
Name: shipping_price, Length: 54639, dtype: float64


In [14]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax = 0.0925

df['total_price'] = df['subtotal'] + df['shipping_price']
df['price_with_tax'] = df['subtotal'] + df['shipping_price'] * (1 + sales_tax)

In [19]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost_each_line'] = df['unit_price'] * df['qty'] + df['shipping_price']
print(df['cost_each_line'])




0        115239.00
1           538.85
2           544.08
3          1068.48
4          4366.30
           ...    
54634      2765.79
54635      9826.73
54636     31230.10
54637       346.16
54638      3124.01
Name: cost_each_line, Length: 54639, dtype: float64


In [18]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['unit_price'] - df['unit_cost']
print(df['line_profit'])



0        334.09
1          9.86
2          5.66
3         11.57
4         86.93
          ...  
54634     31.53
54635     31.13
54636      8.35
54637      0.26
54638      1.40
Name: line_profit, Length: 54639, dtype: float64


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [23]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]

ids_specific = df[df['order_id'].isin(order_ids)]

total_prices = df.groupby('order_id')['unit_price'].sum()

print("Calculated Total Prices:")
print(total_prices)

Calculated Total Prices:
order_id
1000886     523.59
1001431    1094.37
1003003     165.13
1003666    1332.37
1004537     878.13
            ...   
9995243    1208.19
9995530     105.52
9997244     692.18
9997426       4.22
9998480     621.45
Name: unit_price, Length: 10000, dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [25]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
df['total_spent'] = df['qty'] * df['unit_price'] + df['shipping_price']

client_spending = df.groupby('client_id')['total_spent'].sum()

highest_spending_clients = client_spending.sort_values(ascending=False).head()

print("Top 5 Clients by Spending:")
print(highest_spending_clients)



Top 5 Clients by Spending:
client_id
78965    5.950544e+08
85931    2.071955e+08
38473    1.636283e+08
86721    1.015892e+08
84344    9.287877e+07
Name: total_spent, dtype: float64


In [26]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
df['total_spent'] = df['qty'] * df['unit_price'] + df['shipping_price']
df['total_profit'] = df['qty'] * df['unit_price'] - df['qty'] * df['unit_price']

client_summary = df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_spent': 'sum',
    'total_profit': 'sum'
})

highest_spending_clients = client_summary.sort_values(by='total_spent', ascending=False).head(5)

print("Summary DataFrame for Top 5 Clients:")
print(highest_spending_clients)

Summary DataFrame for Top 5 Clients:
               qty  shipping_price   total_spent  total_profit
client_id                                                     
78965      4033302          7383.3  5.950544e+08           0.0
85931       644423          2763.0  2.071955e+08           0.0
38473       402770          7832.2  1.636283e+08           0.0
86721      1010426          6154.6  1.015892e+08           0.0
84344       466968          6944.0  9.287877e+07           0.0


In [27]:
# Format the data and rename the columns to names suitable for presentation.
highest_spending_clients = highest_spending_clients.rename(columns={
    'qty': 'Total Units Purchased',
    'shipping_price': 'Total Shipping Price',
    'total_spent': 'Total Revenue',
    'total_profit': 'Total Profit'
})

highest_spending_clients['Total Units Purchased'] = highest_spending_clients['Total Units Purchased'].astype(int)
highest_spending_clients['Total Shipping Price'] = highest_spending_clients['Total Shipping Price'].map('${:,.2f}'.format)
highest_spending_clients['Total Revenue'] = highest_spending_clients['Total Revenue'].map('${:,.2f}'.format)
highest_spending_clients['Total Profit'] = highest_spending_clients['Total Profit'].map('${:,.2f}'.format)
# Define the money columns. 
money_columns = ['total_revenue', 'total_cost', 'total_profit']

# Define a function that converts a dollar amount to millions.
def dollars_in_millions(amount):
    return amount / 1_000_000.0
# Apply the currency_format_millions function to only the money columns. 
df[money_columns] = df[money_columns].applymap(dollars_in_millions)

# Rename the columns to reflect the change in the money format.
new_column_names = {
    'total_revenue': 'Total Revenue (Millions)',
    'total_cost': 'Total Cost (Millions)',
    'total_profit': 'Total Profit (Millions)'
}
df = df.rename(columns=new_column_names)

print(df)




KeyError: "['total_revenue', 'total_cost'] not in index"

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995